### PR TITLE
Corrects thin ice/snow treatment of enthalpy and other tracers

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Corrects thin ice/snow treatment of enthalpy and other tracers
- [X] Developer(s): 
    njeffery
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Icepack is bit-for-bit, CICE is not.  https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#7e0821dcfe824ae99a3bf0cb957497a8e10c4da7
https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#e8a69abde90b99fc6528d469b8698506a99f6e2a
- How much do the PR code changes differ from the unmodified code? 
    - [ ] bit for bit
    - [ ] different at roundoff level
    - [X] more substantial for CICE?
- Does this PR create or have dependencies on CICE or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please provide any additional information or relevant details below:

Adopted from https://github.com/E3SM-Project/E3SM/pull/5630

"This fix redistributes enthalpy and other tracers evenly in snow and ice when their respective thicknesses are < 1e-15 . Previously, these tracers were zero'd non-conservatively. Also corrects a bug in the zeroing of snow thicknesses, and removes snow in thickness_changes if the ice vanishes."

### CODE REVIEW

File: columnphysics/icepack_therm_vertical.F90 [hunk 1]
```diff
@@ -1706,17 +1706,29 @@ subroutine thickness_changes (nilyr,     nslyr,    &
...
 
       if (ktherm == 2) then
-         if (hsn <= puny) then
+         if (hsn <= puny .or. hin <= c0) then
             do k = 1, nslyr
                fhocnn = fhocnn &
                       + zqsn(k)*hsn/(real(nslyr,kind=dbl_kind)*dt)
                zqsn(k) = -rhos*Lfresh
+!tcx, tcraig, in columnphysics, this is 
+! is it correct that now everything is "if snwgrain"?
```
Comment (eclare108213) [inline columnphysics/icepack_therm_vertical.F90:1715]: Yes, I think `snwgrain` is the correct wrapper for these variables.

**Comment by apcraig** [comment]
Test results look fine.  Everything is running, Icepack is bit-for-bit, CICE is not.

https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#7e0821dcfe824ae99a3bf0cb957497a8e10c4da7

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#e8a69abde90b99fc6528d469b8698506a99f6e2a

**Review by dabail10 — APPROVED** [review]

**Comment by eclare108213** [comment]
Is it worth running the QC tests on this, since it's not BFB?  I think it's fine without them since the cases when the changes occur ought to be rare, but it would provide some extra confidence.

**Comment by apcraig** [comment]
Good idea about QC.  QC passes with very small differences (this is a standard test with snwgrain off)

```
INFO:__main__:Running QC test on the following directories:
INFO:__main__:  /glade/scratch/tcraig/CICE_RUNS/cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.cne03
INFO:__main__:  /glade/scratch/tcraig/CICE_RUNS/cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.T230826b
INFO:__main__:Number of files: 1825
INFO:__main__:2 Stage Test Passed
INFO:__main__:Quadratic Skill Test Passed for Northern Hemisphere
INFO:__main__:Quadratic Skill Test Passed for Southern Hemisphere
INFO:__main__:Creating map of the data (ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.cne03.png)
INFO:__main__:Creating map of the data (ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.T230826b.png)
INFO:__main__:Creating map of the data (ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.cne03_minus_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk.T230826b.png)
INFO:__main__:
INFO:__main__:Quality Control Test PASSED
```

ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk cne03
ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk T230826b
ice_thickness_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk cne03_minus_cheyenne_intel_qcchk_gx1_144x1_medium_qc_qcchk T230826b

**Review by eclare108213 — APPROVED** [review]
Perfect, thanks for running the QC test.

**Pull request merged**